### PR TITLE
SONAR-6472 New WS to create user groups

### PR DESCRIPTION
**Pull request by jblievremont**

### CODE REVIEW

**Comment by SonarTech** [comment]
SonarQube analysis reported 2 new issues:
* 1 major
* 1 info

Details:
* Remove this empty statement. (org.codehaus.sonar:sonar-core:src/main/java/org/sonar/core/user/GroupMapper.java) [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AEmptyStatementUsageCheck)
* Complete the task associated to this TODO comment. (org.codehaus.sonar:sonar-server:src/main/java/org/sonar/server/usergroups/ws/CreateAction.java) [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS1135)

File: server/sonar-server/src/main/java/org/sonar/server/platform/platformlevel/PlatformLevel4.java [hunk 1]
```diff
@@ -493,6 +493,7 @@ public PlatformLevel configure() {
       GroupMembershipFinder.class,
       UserGroupsWs.class,
       org.sonar.server.usergroups.ws.SearchAction.class,
+      org.sonar.server.usergroups.ws.CreateAction.class,
```
Comment (teryk) [inline server/sonar-server/src/main/java/org/sonar/server/platform/platformlevel/PlatformLevel4.java]: please never forget to add component to platformlevel4.java : )

File: server/sonar-server/src/main/java/org/sonar/server/usergroups/ws/CreateAction.java [hunk 2]
```diff
@@ -0,0 +1,129 @@
...
+      .setRequired(true);
+
+    action.createParam(PARAM_DESCRIPTION)
+      .setDescription("Description for the new group. A group description cannot be larger than 200 characters.")
+      .setExampleValue("Default group for new users");
+  }
+
+  @Override
+  public void handle(Request request, Response response) throws Exception {
+    userSession.checkLoggedIn().checkGlobalPermission(GlobalPermissions.SYSTEM_ADMIN);
```
Comment (teryk) [inline server/sonar-server/src/main/java/org/sonar/server/usergroups/ws/CreateAction.java]: why check logged in and global permission at the same time ?

File: server/sonar-server/src/main/java/org/sonar/server/usergroups/ws/CreateAction.java [hunk 3]
```diff
@@ -0,0 +1,129 @@
...
+    String name = request.mandatoryParam(PARAM_NAME);
+    String description = request.param(PARAM_DESCRIPTION);
+
+    validateName(name);
+    validateDescription(description);
+
+    GroupDto newGroup = new GroupDto().setName(name).setDescription(description);
+    DbSession session = dbClient.openSession(false);
+    try {
+      checkNameIsUnique(name, session);
```
Comment (teryk) [inline server/sonar-server/src/main/java/org/sonar/server/usergroups/ws/CreateAction.java]: is there a unique index on the name ?

File: server/sonar-server/src/main/java/org/sonar/server/usergroups/ws/CreateAction.java [hunk 4]
```diff
@@ -0,0 +1,129 @@
...
+      session.commit();
+    } finally {
+      closeQuietly(session);
+    }
+
+    response.newJsonWriter().beginObject().name("group").beginObject()
+      .prop("id", newGroup.getId().toString())
+      .prop(PARAM_NAME, newGroup.getName())
+      .prop(PARAM_DESCRIPTION, newGroup.getDescription())
+      .prop("membersCount", 0)
```
Comment (teryk) [inline server/sonar-server/src/main/java/org/sonar/server/usergroups/ws/CreateAction.java]: why this param ?

File: server/sonar-server/src/main/java/org/sonar/server/usergroups/ws/CreateAction.java [hunk 5]
```diff
@@ -0,0 +1,129 @@
...
+      checkNameIsUnique(name, session);
+      newGroup = dbClient.groupDao().insert(session, new GroupDto().setName(name).setDescription(description));
+      session.commit();
+    } finally {
+      closeQuietly(session);
+    }
+
+    response.newJsonWriter().beginObject().name("group").beginObject()
+      .prop("id", newGroup.getId().toString())
+      .prop(PARAM_NAME, newGroup.getName())
```
Comment (teryk) [inline server/sonar-server/src/main/java/org/sonar/server/usergroups/ws/CreateAction.java]: param and field ?

File: server/sonar-server/src/main/java/org/sonar/server/usergroups/ws/CreateAction.java (hunk 2 above)
Comment (jblievremont) [inline server/sonar-server/src/main/java/org/sonar/server/usergroups/ws/CreateAction.java]: I always do it like this: ``checkLoggedIn`` yields a 401 if unauthenticated, ``checkGlobalPermission`` a 403 if not authorized.

File: server/sonar-server/src/main/java/org/sonar/server/usergroups/ws/CreateAction.java (hunk 3 above)
Comment (jblievremont) [inline server/sonar-server/src/main/java/org/sonar/server/usergroups/ws/CreateAction.java]: No... Actually there is no index on ``groups`` table (except the primary key).

File: server/sonar-server/src/main/java/org/sonar/server/usergroups/ws/CreateAction.java (hunk 4 above)
Comment (jblievremont) [inline server/sonar-server/src/main/java/org/sonar/server/usergroups/ws/CreateAction.java]: To be consistent with the ``search`` action.

File: server/sonar-server/src/main/java/org/sonar/server/usergroups/ws/CreateAction.java (hunk 2 above)
Comment (teryk) [inline server/sonar-server/src/main/java/org/sonar/server/usergroups/ws/CreateAction.java]: I learned something, thanks

File: server/sonar-server/src/main/java/org/sonar/server/usergroups/ws/CreateAction.java (hunk 3 above)
Comment (teryk) [inline server/sonar-server/src/main/java/org/sonar/server/usergroups/ws/CreateAction.java]: shouldn't we add a unique index on the name column ?

File: server/sonar-server/src/main/java/org/sonar/server/usergroups/ws/CreateAction.java (hunk 4 above)
Comment (teryk) [inline server/sonar-server/src/main/java/org/sonar/server/usergroups/ws/CreateAction.java]: does it make sense to expose this as a SeachAction FIELD constant ?

**Comment by teryk** [comment]
add a unique index and LGTM

**Pull request closed without merging**